site/src/job_queue.rs

```diff
@@ -24,19 +27,118 @@ async fn enqueue_master_commits(ctxt: &Arc<SiteCtxt>) {
+                BenchmarkRequestStatus::ArtifactsReady,
                 "",
                 "",
             );
             conn.insert_benchmark_request(&benchmark).await;
         }
     }
+    Ok(())
+}
+
+/// This function requires the sorts the priority order of the queue. It assumes
+/// that the only status is `ArtifactsReady`
+fn sort_benchmark_requests<'a>(
+    done: &mut HashSet<String>,
+    unordered_queue: &'a mut [BenchmarkRequest],
+) -> &'a [BenchmarkRequest] {
+    // A topological sort, where each "level" is additionally altered such that
+    // try commits come first, and then sorted by PR # (as a rough heuristic for
+    // earlier requests).
+
+    // Ensure the queue is ordered by status and the `commit_type`
+    unordered_queue.sort_unstable_by_key(|bmr| (bmr.status.rank(), bmr.commit_type.rank()));
+
+    let mut finished = 0;
+    while finished < unordered_queue.len() {
+        // The next level is those elements in the unordered queue which
+        // are ready to be benchmarked (i.e., those with parent in done or no
+        // parent).
+        let level_len = partition_in_place(unordered_queue[finished..].iter_mut(), |bmr| {
+            bmr.parent_sha().is_none_or(|parent| done.contains(parent))
+        });
+
+        // No commit is ready for benchmarking. This can happen e.g. when a try parent commit
+        // was forcefully removed from the master branch of rust-lang/rust. In this case, just
+        // let the commits be benchmarked in the current order that we have, these benchmark runs
+        // just won't have a parent result available.
+        if level_len == 0 {
+            return unordered_queue;
+        }
+
+        let level = &mut unordered_queue[finished..][..level_len];
+        level.sort_unstable_by_key(|bmr| {
+            (
+                bmr.parent_sha().is_some(),
+                *bmr.pr().unwrap_or(&0),
+                bmr.created_at,
+                bmr.tag().to_string(),
+            )
+        });
+        for c in level {
+            done.insert(c.tag().to_string());
+        }
+        finished += level_len;
+    }
+    unordered_queue
+}
+
+/// Given some pending requests and a list of completed requests determine if
+/// we have another request the we can process
+fn get_next_benchmark_request<'a>(
+    pending: &'a mut [BenchmarkRequest],
+    completed_set: &mut HashSet<String>,
+) -> Option<&'a BenchmarkRequest> {
+    sort_benchmark_requests(completed_set, pending).first()
+}
+
+/// Enqueue the job into the job_queue
+async fn enqueue_next_job(conn: &mut dyn database::pool::Connection) -> anyhow::Result<()> {
+    let mut pending = conn
+        .get_benchmark_requests_by_status(
+            &[
+                BenchmarkRequestStatus::InProgress,
+                BenchmarkRequestStatus::ArtifactsReady,
+            ],
+            None,
+        )
+        .await?;
+
+    // No requests to process or we have something currently in progress
```

Comment:
The comment is inaccurate, if `pending` is empty, the function will continue below this condition.